File: 2024/dec02/C/aux02.c
#include "dec02.h"

int check_line(int *line, int num) {
    int safe=1; //change to 0 if something is unsafe
    int signal;
    if (line[1]-line[0]>0) signal=1;
    else if (line[1]-line[0]<0) signal=-1;
    else return 0; //if first two values are equal, it's not safe

    for (int i=1; i<num; i++) {
        int delta=line[i]-line[i-1];
        /* Fail if sequence doesn't follow increase/decrease
        or delta isn't a safe value*/
        if (delta*signal<0 || abs(delta)>3 || abs(delta)<1) {
            safe=0;
            break;
        }
    }
    return safe;
}

void remove_line_elem(int *line, int *num, int index) {
    for (int i=index; i<*num-1; i++) {
        line[i]=line[i+1];
    }
    line[*num-1]=0; //last element is 0
    (*num)--; //decrease number of elements
}   

File: 2024/dec02/Python/aux02.py
def check_line_new(seq):
    safe=True
    safeValues=(1,2,3)

    if (seq[1]-seq[0] > 0):
        signal=1
    elif (seq[1]-seq[0] < 0):
        signal=-1
    else:
        return False #if first 2 values are equal, it's not safe
    
    for i in range(1,len(seq)):
        delta=seq[i]-seq[i-1]
        #Fail if sequence doesn't follow increase/decrease
        #or delta isn't a safe value
        if (delta*signal<0 or abs(delta) not in safeValues):
            safe=False
            break

    return safe

def check_line(seq):
    leng = len(seq)
    differences = []
    safeVar=0
    for val in range(0, leng-1):
        differences.append(seq[val+1]-seq[val])  #A lista diferenças contém todas as variações da linha
    signal = 'increase' if differences[0] > 0 else 'decrease'  #O sinal é definido pela primeira variação
    for diff in differences:
            safeVar+=1 if ((signal=='increase' and diff>0) or (signal=='decrease' and diff<0)) and 1<=abs(diff)<=3 else 0
    return True if safeVar==leng-1 else False  #A linha só é seguro se todos os valores forem seguros